contracts/PetPark.sol

```diff
@@ -1,7 +1,93 @@
 //SPDX-License-Identifier: Unlicense
 pragma solidity ^0.8.0;
 
-
 contract PetPark {
 
+    event Added(AnimalType animalType, uint count);
+    event Borrowed(AnimalType animalType);
+    event Returned(AnimalType animalType);
+    
+    enum AnimalType{ None, Fish, Cat, Dog, Rabbit, Parrot }
+    enum Gender{ Male, Female }
+
+    struct Borrower {
+        uint age;
+        Gender gender;
+        AnimalType animal;
+    }
+
+    mapping(address => Borrower) private _borrowers;
+    mapping(AnimalType => uint) private _shelter;
+
+    address private _owner;
+
+    constructor() {
+        _owner = msg.sender;
+    }
+
+    function add(AnimalType animalType, uint count) public onlyOwner {
+        require(existingAnimal(animalType), 'Invalid animal');
+
+        _shelter[animalType] += count;
+
+        emit Added(animalType, count);
+    }
+
+    function borrow(uint age, Gender gender, AnimalType animalType) public {
+        require(age > 0, 'Invalid Age');
+
+        Borrower storage borrower = _borrowers[msg.sender];
+
+        if (borrower.age > 0) {
+            require(borrower.age == age, 'Invalid Age');
+            require(borrower.gender == gender, 'Invalid Gender');
+        }
+
+        require(_borrowers[msg.sender].animal == AnimalType.None, 'Already adopted a pet');
+        
+        require(uint8(animalType) > 0 && uint8(animalType) < 6, 'Invalid animal type');
```

Comment:
You can use the `existingAnimal` function you defined

Reply:
Big thanks for all the comments!